fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/FoldConstantRule.java

```diff
@@ -0,0 +1,440 @@
+    private static final DateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+    private final IdGenerator<ExprId> idGenerator = ExprId.createGenerator();
+
+    private static boolean isAllLiteral(Expression... children) {
+        return Arrays.stream(children).allMatch(c -> c instanceof Literal);
+    }
+
+    private static boolean isAllLiteral(List<Expression> children) {
+        return children.stream().allMatch(c -> c instanceof Literal);
+    }
+
+    private static boolean hasNull(List<Expression> children) {
+        return children.stream().anyMatch(c -> c instanceof NullLiteral);
+    }
+
+    private static boolean hasNull(Expression... children) {
+        return Arrays.stream(children).anyMatch(c -> c instanceof NullLiteral);
+    }
+
+    @Override
+    public Expression rewrite(Expression expr, ExpressionRewriteContext ctx) {
+        Expression expression = expr.accept(this, ctx);
+        if (ctx.connectContext != null && ctx.connectContext.getSessionVariable().isEnableFoldConstantByBe()) {
+            return foldByBe(expression, ctx);
+        }
+        return expression;
+    }
+
+    @Override
+    public Expression visitComparisonPredicate(ComparisonPredicate cp, ExpressionRewriteContext context) {
+        Expression left = rewrite(cp.left(), context);
+        Expression right = rewrite(cp.right(), context);
+
+        if (!(cp instanceof NullSafeEqual) && hasNull(left, right)) {
+            return Literal.of(null);
+        }
+
+        if (isAllLiteral(left, right)) {
+            Literal l = (Literal) left;
+            Literal r = (Literal) right;
+            if (cp instanceof EqualTo) {
+                return BooleanLiteral.of(l.compareTo(r) == 0);
+            }
+            if (cp instanceof GreaterThan) {
+                return BooleanLiteral.of(l.compareTo(r) > 0);
+            }
+            if (cp instanceof GreaterThanEqual) {
+                return BooleanLiteral.of(l.compareTo(r) >= 0);
+            }
+            if (cp instanceof LessThan) {
+                return BooleanLiteral.of(l.compareTo(r) < 0);
+            }
+            if (cp instanceof LessThanEqual) {
+                return BooleanLiteral.of(l.compareTo(r) <= 0);
+            }
+            if (cp instanceof NullSafeEqual) {
+                if (l.isNull() && r.isNull()) {
+                    return BooleanLiteral.TRUE;
+                } else if (!l.isNull() && !r.isNull()) {
+                    return BooleanLiteral.of(l.compareTo(r) == 0);
+                } else {
+                    return BooleanLiteral.FALSE;
+                }
+            }
+        }
+        return cp.withChildren(left, right);
+    }
+
+    @Override
+    public Expression visitNot(Not not, ExpressionRewriteContext context) {
+        Expression child = rewrite(not.child(), context);
+        if (child instanceof NullLiteral) {
+            return Literal.of(null);
+        }
```

Comment:
this part look like a common logic, you can move to the common visit method, like this:
```java
public Expression visit(Expression expr, ExpressionRewriteContext context) {
  List<Expression> children = expr.children()
         .stream()
         .map(child -> child.visit(this, context)
         .collect(Collectors.toList());
    boolean hasNullLiteralChild = ExpresstionUtils.hasNullLiteral(children);

    // isDeterminate: Env.getCurrentEnv().isNullResultWithOneNullParamFunction()
    if (expr.isDeterminate() && hasNullLiteralChild) {
      return expr;
    }

    if (!ExpressionUtils.isAllLiteral(children) {
      return expr;
    }
    // this position every child is Literal, and forward to visitXxx method
    return expr.withChildren(children).accept(this, context);
}
```

so every visitXxx method can simple fold constant because every child must be Literal